### PR TITLE
Ensure dicts are properly recognized as msgpack serializable

### PR DESCRIPTION
**Pull request by fjetter**
This was already introduces in https://github.com/dask/distributed/pull/3536 pretty sure this is a typo.

This typo effectively prohibited simple dicts without any custom objects to take the fast path since `all(typ is str for x in v.keys())` is always False

similarly, genuine bytes where also misclassified and encoded using a "complex" header.

I doubt that either is a big deal but since this was introduced for cuda I guess smbd from @dask/gpu should be aware.

About the test... ObjectDictSerializer is not tested explicitly anywhere and I chose not to introduce a granular unit test for this, yet. I'm not super happy about the `test__is_msgpack_serializable` but at least for the composite/nested types this adds _some_ value

### CODE REVIEW

File: distributed/protocol/serialize.py [hunk 1]
```diff
@@ -817,11 +817,12 @@ def _is_msgpack_serializable(v):
         v is None
         or typ is str
         or typ is bool
+        or typ is bytes
         or typ is int
         or typ is float
         or isinstance(v, dict)
         and all(map(_is_msgpack_serializable, v.values()))
-        and all(typ is str for x in v.keys())
+        and all(isinstance(x, str) for x in v.keys())
```
Comment (graingert) [inline distributed/protocol/serialize.py:825]: is there a reason for not opting for:

```suggestion
        and all(type(x) is str for x in v.keys())
```

Comment (fjetter) [reply distributed/protocol/serialize.py:825]: no particular reason

Comment (graingert) [reply distributed/protocol/serialize.py]: with `strict_types` it's not possible to serialize str subclasses

```python
>>> class Foo(str): pass
... 
>>> msgpack.dumps({Foo("x"): "y"}, use_bin_type=True, strict_types=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/graingert/anaconda3/envs/dask-distributed/lib/python3.10/site-packages/msgpack/__init__.py", line 38, in packb
    return Packer(**kwargs).pack(o)
  File "msgpack/_packer.pyx", line 294, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 300, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 297, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 229, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 291, in msgpack._cmsgpack.Packer._pack
TypeError: can not serialize 'Foo' object
>>> 
```

**Review by graingert — APPROVED** [review]

**Review by wence- — APPROVED** [review]

**Review by jrbourbeau — COMMENTED** [review]
Thanks for fixing @fjetter and reviewing @graingert @wence-!

**Pull request merged**